lib/componentmanager.js

```diff
@@ -429,28 +473,56 @@
      * 
      * @see ComponentManager.prototype.addComponent
      * @param {Layer} layer The layer from which to extract components
-     * @return {Array.<{errors: Array.<string>} | {component: Component}>}
+     * @return {Promise} resolves to Array.<{errors: Array.<string>} | {component: Component}>}
      */
     ComponentManager.prototype.findAllComponents = function (layer) {
         
-        var results,
-            components = [];
-
-        if (layer.name) {
-            results = this._parserManager.analyzeLayerName(layer.name);
-            results.forEach(function (result) {
-                var errors = result.errors;
-                if (errors.length > 0) {
-                    components.push({errors: errors});
-                } else {
-                    var component = result.component;
-                    if (component.file || component.default) {
-                        components.push({component: component});
-                    }
+        var findAllDeferred = Q.defer(),
+            _G = this._generator,
+            _PM = this._parserManager,
+            doc = this._document,
+            fnFromLayerName = function (layerName) {
+                var aRet = [],
+                    results;
+                if (layerName) {
+                    results = _PM.analyzeLayerName(layerName);
+                    results.forEach(function (result) {
+                        var errors = result.errors;
+                        if (errors.length > 0) {
+                            aRet.push({errors: errors});
+                        } else {
+                            var component = result.component;
+                            if (component.file || component.default) {
+                                aRet.push({component: component});
+                            }
+                        }
+                    });
                 }
-            }, this);
+                return aRet;
+            };
+        
+        if (!this._config["meta-data-driven"]) {
+            findAllDeferred.resolve(fnFromLayerName(layer.name));
+        } else {
+            _G.getDocumentSettingsForPlugin(doc.id, META_PLUGIN_ID).then(function (docMeta) {
+                if (docMeta && docMeta.metaEnabled) {
+                    _G.getLayerSettingsForPlugin(doc.id, layer.id, META_PLUGIN_ID).then(function (layerMeta) {
+                        var components = [];
+                        if (layerMeta && layerMeta.assetSettings) {
+                            layerMeta.assetSettings.forEach(function (setting) {
+
+                                //TBD: translation, if desired
+
+                                components.push({component: setting});
+                            }.bind(this));
+                        }
+                        findAllDeferred.resolve(components);
```

Comment:
instead of having a deferred handing around I think it would be safer if it did something like ```finalPromise = _G.getDocumentSettingsForPlugin)...```. The promise is the result from the ```then()```. At this spot here just ```return components; ``` and the components will get passed back in the resolved promise. 

The main code issue here is the else case doesn't resolve or reject the promise so if it ever happened the code would just be waiting on a promise that never resolved or rejected.